src/runtime/timer.c

```diff
@@ -6,136 +6,74 @@
-    clock_id id;
-    boolean disable;
-};
-
 // should pass a timer around
 static pqueue timers;
 static heap theap;
 
-#if defined(STAGE3) || defined(BUILD_VDSO)
-#include <vdso.h>
-#define rtc_offset (&(VVAR_REF(vdso_dat)))->rtc_offset
-#else
-#define rtc_offset 0
-#endif
-
-static inline timestamp expiry(timer t)
-{
-    switch (t->id) {
-    case CLOCK_ID_MONOTONIC:
-    case CLOCK_ID_MONOTONIC_RAW:
-    case CLOCK_ID_MONOTONIC_COARSE:
-    case CLOCK_ID_BOOTTIME:
-        return t->expiry;
-    case CLOCK_ID_REALTIME:
-    case CLOCK_ID_REALTIME_COARSE:
-        return t->expiry - rtc_offset;
-    default:
-        halt("expiry: clock id %d unsupported\n");
-    }
-}
-
 /* The lower time expiry is the higher priority. */
 static boolean timer_compare(void *za, void *zb)
 {
-    return expiry((timer)za) > expiry((timer)zb);
+    return timer_expiry((timer)za) > timer_expiry((timer)zb);
 }
 
-/* returns time remaining or 0 if elapsed */
-timestamp remove_timer(timer t)
+define_closure_function(1, 0, void, timer_free,
+                        timer, t)
 {
-    t->disable = true;
-    timestamp x = expiry(t);
-    timestamp n = now(CLOCK_ID_MONOTONIC);
-    return x > n ? x - n : 0;
+    deallocate(theap, bound(t), sizeof(struct timer));
 }
 
-static timer __register_timer(timestamp interval, clock_id id, thunk n, boolean periodic)
+timer register_timer(clock_id id, timestamp val, boolean absolute, timestamp interval, timer_handler n)
 {
-    timer t=(timer)allocate(theap, sizeof(struct timer));
+    timer t = allocate(theap, sizeof(struct timer));
     if (t == INVALID_ADDRESS) {
         msg_err("failed to allocate timer\n");
         return INVALID_ADDRESS;
     }
 
-    timestamp tn = now(id);
-    t->t = n;
-    t->disable = false;
-    t->expiry = tn + interval;
-    t->interval = (periodic) ? interval : 0;
     t->id = id;
-    pqueue_insert(timers, t);
-
-    timer_debug("register %s timer: %p, interval %T, now %T, expiry %T\n",
-                (periodic) ? "periodic" : "one-shot", t, interval, tn, t->expiry);
-
-    return(t);
-}
+    t->expiry = absolute ? val : now(id) + val;
+    t->interval = interval;
+    t->disabled = false;
+    t->t = n;
 
-timer register_timer(timestamp interval, clock_id id, thunk n)
-{
-    return __register_timer(interval, id, n, false);
+    init_refcount(&t->refcount, 1, init_closure(&t->free, timer_free, t));
+    pqueue_insert(timers, t);
+    timer_debug("register timer: %p, expiry %T, interval %T, handler %p\n", t, t->expiry, interval, n);
+    return t;
 }
 
-timer register_periodic_timer(timestamp interval, clock_id id, thunk n)
+timestamp timer_check(void)
 {
-    return __register_timer(interval, id, n, true);
-}
-    
-/* Presently called with ints off. Address thread safety with
-   pqueue before using with ints enabled.
-*/
-timestamp timer_check()
-{
-    timestamp here = 0;
-    timer t = 0;
+    timer t;
+    s64 delta;
+    timestamp here = now(CLOCK_ID_MONOTONIC);
```

Comment:
its a bit fussy, and I don't know the details. but is CLOCK_ID_MONOTIC really something that's fundamental or is it semantics carried from the linux interface?

Reply:
I believe it's from posix 1b. To support timer_create, clock_nanosleep, etc, we need to somehow parameterize the clock source used as a reference. Not clear that this is the best way to do it; we could have instead used a vft per clock id.